rustworkx-core/src/generators/configuration_model.rs

```diff
@@ -0,0 +1,344 @@
+// under the License.
+
+use crate::generators::InvalidInputError;
+use num_traits::{Num, ToPrimitive};
+use petgraph::data::{Build, Create};
+use petgraph::visit::{Data, GraphProp, NodeIndexable};
+use petgraph::Undirected;
+use rand::distributions::Distribution;
+use rand::prelude::SliceRandom;
+use rand::Rng;
+use std::iter::repeat;
+
+/// A degree sequence is a vector of integers that sum up to an even number.
+///
+/// A graph with a given degree sequence can be generated with
+/// [configuration_model](fn.configuration_model.html).
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct DegreeSequence {
+    values: Vec<usize>,
+}
+
+impl DegreeSequence {
+    /// Returns a new degree sequence.
+    ///
+    /// Arguments:
+    /// * `vec`: A vector of integers that sum up to an even number.
+    pub fn new(vec: Vec<usize>) -> Result<Self, DegreeSequenceError> {
+        let sum: usize = vec.iter().sum();
+        if sum % 2 != 0 {
+            return Err(DegreeSequenceError::OddSum);
+        }
+        Ok(Self { values: vec })
+    }
+
+    /// Generates a random degree sequence with the given length and values
+    /// sampled from the given distribution.
+    ///
+    /// Arguments:
+    /// * `rng`: A random number generator.
+    /// * `length`: The length of the degree sequence.
+    /// * `distribution`: A distribution that samples values that are convertible to `usize`.
+    ///
+    /// NOTE: If the sampled values are floating point numbers, they will be truncated.
+    ///
+    /// # Example
+    /// ```rust
+    /// use rand::SeedableRng;
+    /// use rustworkx_core::petgraph::graph::UnGraph;
+    /// use rustworkx_core::generators::DegreeSequence;
+    /// use rand_distr::Poisson;
+    ///
+    /// let mut rng = rand::rngs::StdRng::seed_from_u64(42);
+    /// let poisson = Poisson::new(2.0).unwrap();
+    /// let degree_sequence = DegreeSequence::from_distribution(&mut rng, 10, &poisson).unwrap();
+    /// let expected_sequence = vec![3, 0, 4, 0, 3, 1, 0, 1, 4, 2];
+    /// assert_eq!(*degree_sequence, expected_sequence);
+    /// ```
+    pub fn from_distribution<R: Rng + ?Sized, I: Num + PartialOrd + ToPrimitive>(
+        rng: &mut R,
+        length: usize,
+        distribution: &impl Distribution<I>,
+    ) -> Result<Self, DegreeSequenceError> {
+        let mut degree_sequence: Vec<usize> = Vec::with_capacity(length);
+        let mut sum: usize = 0;
+        for value in distribution.sample_iter(&mut *rng).take(length) {
+            if value < I::zero() {
+                return Err(DegreeSequenceError::NegativeDegree);
+            }
+            let value = match value.to_usize() {
+                Some(value) => value,
+                None => return Err(DegreeSequenceError::Conversion),
+            };
+            degree_sequence.push(value);
+            sum += value;
+        }
+        if sum % 2 == 1 {
+            let index = rng.gen_range(0..length);
+            degree_sequence[index] += 1;
```

Comment:
This felt like a "pragmatic" way to ensuring that the sum is even. However, this could lead to out-of-distribution values in the sequence which is probably confusing for the user. Not sure how to fix it. Probably just retry n times?